SolutionArray formatting tests fail with fmt 6.1.2

<!-- Please fill in the following information to report a problem with Cantera. If you have a question about using Cantera, please post it on our Google Users' Group (https://groups.google.com/forum/#!forum/cantera-users). Feature enhancements should be discussed in the dedicated Cantera enhancements repository (https://github.com/Cantera/enhancements/new/choose) -->

**Problem description**

Using `fmt` version 6.1.2, a number of tests in the Python test suite that test the formatted output from `SolutionArray` report failures. This version of `fmt` is what is available on Ubuntu 20.04, so it would be nice to continue supporting it.

**Steps to reproduce**

<!-- A minimal, complete, and reproducible example demonstrating the problem. See https://stackoverflow.com/help/minimal-reproducible-example for additional suggestions on how to create such an example. -->

1. Compile Cantera using a system copy of `fmt` version 6.1.2
2. Run `scons test-python-composite`

**Behavior**

The following tests fail:
```
============================================== short test summary info ===============================================
XFAIL test/python/test_composite.py::TestLegacyHDF::test_legacy_hdf_str_column - Unable to read fixed length strings from HDF
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_double_vector - AssertionError: assert 77 == 79
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_integer_vector - AssertionError: assert 79 == 81
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_long - AssertionError: assert 74 == 76
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_plus_minus_e - AssertionError: assert 79 == 81
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_plus_minus_f - AssertionError: assert 79 == 81
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_plus_minus_i - AssertionError: assert 73 == 77
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_scientific - AssertionError: assert 68 == 69
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_select_rows - AssertionError: assert 68 == 70
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_short - AssertionError: assert 68 == 70
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_string_vector - AssertionError: assert 77 == 79
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_strings - AssertionError: assert 76 == 78
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_water_extra - AssertionError: assert 22 == 24
FAILED test/python/test_composite.py::TestSolutionArrayInfo::test_water_simple - AssertionError: assert 16 == 18
====================================== 13 failed, 81 passed, 1 xfailed in 1.66s ======================================
```

A specific error that helps show that the output is indeed not as expected:
```
_______________________________________ TestSolutionArrayInfo.test_select_rows _______________________________________

self = <python.test_composite.TestSolutionArrayInfo testMethod=test_select_rows>

    def test_select_rows(self):
        arr = ct.SolutionArray(self.gas, 25)
        ix = [2, 5, 6, 9, 15, 3]
        arr2 = arr[ix]
>       self.check(arr2, arr2.info(width=self.width), 6)

test/python/test_composite.py:380: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <python.test_composite.TestSolutionArrayInfo testMethod=test_select_rows>
arr =       T          D   H2    H    O   O2   OH  H2O  HO2 H2O2   AR   N2
2.0 300.0  0.0818939  1.0  0.0  0.0  0.0  0.0  0.... 0.0  0.0
3.0 300.0  0.0818939  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0

[6 rows x 12 components; state='TDY']
repr = "      T          D   H2    H    O   O2   OH  H2O  HO2 H2O2   AR   N2\n2.0 300.0  0.0818939  1.0  0.0  0.0  0.0  0.0  ...  0.0\n3.0 300.0  0.0818939  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n\n[6 rows x 12 components; state='TDY']"
rows = 6

    def check(self, arr, repr, rows):
        count = 0
        width = None
        header = None
        for line in repr.split("\n"):
            if not len(line):
                break
            if width is None:
                width = len(line)
                assert width <= self.width
                header = line.split()
            else:
>               assert width == len(line)
E               AssertionError: assert 68 == 70
E                +  where 70 = len('2.0 300.0  0.0818939  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0')

test/python/test_composite.py:305: AssertionError
```
You can see that in the first data row, there is an ellipsis, not separated from adjacent entries, and also not present in either the header row or the following data row.

**System information**

- Cantera version: 3.0.0b1
- OS: Ubuntu 20.04
- Python/MATLAB/other software versions:
  - Python 3.8.10
  - fmt 6.1.2

**Attachments**

Complete log of running `scons test-python-composite`: [test-python-composite-log.txt](https://github.com/Cantera/cantera/files/11952311/test-python-composite-log.txt)

Finally managed to get a configuration on macOS (Python 3.8 / fmt 6.1.2), by simply checking out the 6.1.2 hash from the submodule. It does, however, not compile with the following error right at the beginning:

For `main`:

```
g++ -o build/src/extensions/pythonShim.os -c -isystem /opt/homebrew/Caskroom/miniforge/base/envs/ctdev/lib/python3.8/site-packages/numpy/core/include -isystem /opt/homebrew/Caskroom/miniforge/base/envs/ctdev/include/python3.8 -isystem include/cantera/ext -isystem /opt/homebrew/Caskroom/miniforge/base/envs/ctdev/include -std=c++17 -Wno-deprecated-declarations -O3 -Wno-inline -g -Wall -fPIC -DNDEBUG -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -Iinclude -Iinclude -Ibuild/src src/extensions/pythonShim.cpp
scons: *** [build/ext/fmt/src/os.os] Source `ext/fmt/src/os.cc' not found, needed by target `build/ext/fmt/src/os.os'.
```

For the #1538 branch:
```
g++ -o build/src/extensions/pythonShim.os -c -isystem /opt/homebrew/Caskroom/miniforge/base/envs/ctdev/lib/python3.8/site-packages/numpy/core/include -isystem /opt/homebrew/Caskroom/miniforge/base/envs/ctdev/include/python3.8 -isystem include/cantera/ext -isystem /opt/homebrew/Caskroom/miniforge/base/envs/ctdev/include -std=c++17 -Wno-deprecated-declarations -O3 -Wno-inline -g -Wall -fPIC -DNDEBUG -DFMT_HEADER_ONLY=1 -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -Iinclude -Iinclude -Ibuild/src src/extensions/pythonShim.cpp
ext/fmt/src/format.cc:124:35: error: duplicate explicit instantiation of 'basic_data<>'
template struct FMT_API internal::basic_data<void>;
                                  ^
ext/fmt/include/fmt/format.h:739:28: note: previous explicit instantiation is here
FMT_EXTERN template struct basic_data<void>;
                           ^
1 error generated.
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -45,7 +45,7 @@ jobs:
       run: |
         sudo apt update
         sudo apt install libboost-dev gfortran libopenmpi-dev libpython3-dev \
-        libblas-dev liblapack-dev libhdf5-dev
+        libblas-dev liblapack-dev libhdf5-dev libfmt-dev
         gcc --version
     - name: Upgrade pip
       run: python3 -m pip install -U pip setuptools wheel
@@ -58,7 +58,7 @@ jobs:
     - name: Build Cantera
       run: |
         python3 `which scons` build env_vars=all -j2 debug=n --debug=time \
-        hdf_libdir=$HDF5_LIBDIR hdf_include=$HDF5_INCLUDEDIR \
+        system_fmt=y hdf_libdir=$HDF5_LIBDIR hdf_include=$HDF5_INCLUDEDIR \
         cc_flags=-D_GLIBCXX_ASSERTIONS
     - name: Upload shared library
       uses: actions/upload-artifact@v3
@@ -801,7 +801,7 @@ jobs:
       run: brew install --display-times hdf5
       if: matrix.os == 'macos-11'
     - name: Install Apt dependencies (Ubuntu)
-      run: sudo apt install libhdf5-dev
+      run: sudo apt install libhdf5-dev libfmt-dev
       if: matrix.os == 'ubuntu-22.04'
     - name: Setup .NET Core SDK
       uses: actions/setup-dotnet@v2

--- a/src/base/SolutionArray.cpp
+++ b/src/base/SolutionArray.cpp
@@ -295,7 +295,7 @@ vector<string> integerColumn(string name, const vector<long int>& comp,
                              int rows, int width)
 {
     // extract data for processing
-    vector<double> data;
+    vector<long int> data;
     string notation = fmt::format("{{:{}}}", width);
     size_t maxLen = 2; // minimum column width is 2
     int csize = static_cast<int>(comp.size());

--- a/src/base/SolutionArray.cpp
+++ b/src/base/SolutionArray.cpp
@@ -273,7 +273,7 @@ vector<string> doubleColumn(string name, const vector<double>& comp,
         notation = fmt::format(" {{:>{}.{}f}}", over + maxLen, tail);
     } else {
         // all entries are integers
-        notation = fmt::format(" {{:>{}}}", over + maxLen);
+        notation = fmt::format(" {{:>{}.0f}}", over + maxLen);
     }
     maxLen = fmt::format(notation, 0.).size();
 

--- a/src/base/SolutionArray.cpp
+++ b/src/base/SolutionArray.cpp
@@ -303,29 +303,29 @@ vector<string> integerColumn(string name, const vector<long int>& comp,
     if (csize <= rows) {
         for (const auto& val : comp) {
             data.push_back(val);
-            string name = boost::trim_copy(fmt::format(notation, val));
-            if (name[0] == '-') {
-                name = name.substr(1);
+            string formatted = boost::trim_copy(fmt::format(notation, val));
+            if (formatted[0] == '-') {
+                formatted = formatted.substr(1);
             }
-            maxLen = std::max(maxLen, name.size());
+            maxLen = std::max(maxLen, formatted.size());
         }
     } else {
         dots = (rows + 1) / 2;
         for (int row = 0; row < dots; row++) {
             data.push_back(comp[row]);
-            string name = boost::trim_copy(fmt::format(notation, comp[row]));
-            if (name[0] == '-') {
-                name = name.substr(1);
+            string formatted = boost::trim_copy(fmt::format(notation, comp[row]));
+            if (formatted[0] == '-') {
+                formatted = formatted.substr(1);
             }
-            maxLen = std::max(maxLen, name.size());
+            maxLen = std::max(maxLen, formatted.size());
         }
         for (int row = csize - rows / 2; row < csize; row++) {
             data.push_back(comp[row]);
-            string name = boost::trim_copy(fmt::format(notation, comp[row]));
-            if (name[0] == '-') {
-                name = name.substr(1);
+            string formatted = boost::trim_copy(fmt::format(notation, comp[row]));
+            if (formatted[0] == '-') {
+                formatted = formatted.substr(1);
             }
-            maxLen = std::max(maxLen, name.size());
+            maxLen = std::max(maxLen, formatted.size());
         }
     }
 

